Prep sample plan inputs

gcperk opened the pull request:
This PR add data prep for the landscape scale data layer to generate the cost layer. 
This includes generating the roads, dem and start points for the landscape scale (ie larger than the initial AOI). 

It check roads layer is compliant with coding and checks the start points are attached to the roads network. 

Placeholders are also added for generate cost layer (methods old and new), which will require more testing to assess which will be incorporated into the package (see issue #1). 

Submitting this PR to avoid long length review as continue to work on PEMsamplr

@ateucher - if you want to provide a high level review, as this will require more detailed work for testing/ etc in the future.

gcperk commented:
@ateucher - FYI - this PR includes all steps up to building sample plan. I started a new branch for the build to avoid cross over!

ateucher commented on R/create_bgc_mask.R:
```diff
@@ -0,0 +1,98 @@
+      "bec.gpkg does not exist in {.var vec_dir}. Please check the function
+        create_base_vectors() ran correctly or add this manually"
+    )
+  } else {
```
Since the previous block exits if there's an error we don't need to wrap this in `else {}`

```suggestion
  }
```

ateucher commented on R/create_bgc_mask.R:
```diff
@@ -0,0 +1,98 @@
+
+    }
+    invisible(out_dir)
+  }
```
```suggestion

```

ateucher commented on R/create_bgc_mask.R:
```diff
@@ -0,0 +1,98 @@
+  } else {
+    bec <- sf::st_read(fs::path(vec_dir, "bec.gpkg"))
+
+    if (!fs::dir_exists(out_dir)) {
+      fs::dir_create(out_dir, recurse = TRUE)
+      cli::cli_alert_warning(
+        "write out folder does not exist, creating at location {.var out_dir}}"
+      )
+    }
```
I think we can just run this - it creates the dir if it doesn't exist, and is just silent if it does already

```suggestion
  fs::dir_create(out_dir, recurse = TRUE)
```

ateucher commented on R/create_bgc_mask.R:
```diff
@@ -0,0 +1,98 @@
+    for (b in boi) {
+
+      subzone <- bec |>
+        dplyr::filter("MAP_LABEL" %in% b)
```
```suggestion
        dplyr::filter("MAP_LABEL" == b)
```

gcperk replied:
I think originally I had this as %in% as there were some bec variants and subzones which were not captured in the BEC data, so would be combined. I will shift to == but might need changing in the future

ateucher commented on R/create_bgc_mask.R:
```diff
@@ -0,0 +1,98 @@
+    cli::cli_abort("{.var vec_dir} must be a SpatRaster or a path to a file")
+  }
+
+  if (inherits(cost_masked, c("character"))) {
+    cost <- terra::rast(cost_masked)
+  } else if (!inherits(cost_masked, c("SpatRaster"))) {
+    cli::cli_abort("{.var cost_masked} must be a SpatRaster or a path to a file")
+  }
```
This pattern is super common throughout, I wonder if we should make a function like `read_spatrast_if_necessary()` or something which would encapsulate all of this logic?

gcperk replied:
Added a new function to utils on PEMprepr 


`
read_spatrast_if_necessary = function(spatRast){

  if (inherits(spatRast, c("character"))) {
    spatRast <- terra::rast(spatRast)

  } else if (!inherits(spatRast, c("SpatRaster"))) {
    cli::cli_abort("{.var spatRast} must be a SpatRaster or a path to a file")
  }
  return(spatRast)
}

`

ateucher reviewed (approved):
Hey @gcperk! I've left quite a few comments, but most are pretty easily addressed I think. I couldn't really fully assess `check_bgc_cost()`, `create_cost_exclusion()`,  `create_cost_penalty()`, or `create_bgc_mask()` as I couldn't run them without a cost layer (I think they are generated by the placeholders)?

I also didn't really assess the tests, though adding them later I think is important (but may be a big job!)

Nice work! Happy to chat through any of this 😄

ateucher commented:
PS. This is the bulk of the code I ran to review things:

```r
library(PEMprepr)

## PEMprepr setup --------------------------
options(pemprepr.saga_path = "/Applications/QGIS.app/Contents/MacOS/bin/saga_cmd")

make_test_aoi <- function(outdir) {
  PEMprepr:::snap_aoi(
    fs::path_package("PEMprepr", "extdata/datecreek_aoi.gpkg"),
    out_dir = fs::path(outdir, "snap")
  )
}
# saga_cmd()

outdir <- withr::local_tempdir()
aoi <- make_test_aoi(outdir)

aoi_rast <- create_template_raster(aoi, res = 50, out_dir = outdir)

dem <- get_cded_dem(aoi_rast, write_output = FALSE, ask = FALSE)

create_base_vectors(aoi, outdir)
create_covariates(dem, out_dir = outdir)
create_landscape_covariates(dem, out_dir = outdir)
create_bgc_template(
  bec =  sf::read_sf(fs::path(outdir, "bec.gpkg")),
  field = "MAP_LABEL",
  template_rast = aoi_rast,
  write_output = TRUE,
  out_dir = outdir
)

list.files(outdir)

## PEMsamplr testing -----------------------------

debugonce(create_binned_landscape)
combr <- create_binned_landscape(file.path(outdir, "50m", "modules"))

plot(combr)

debugonce(check_bgc_landscapes)
check_bgc_out <- check_bgc_landscapes(
  file.path(outdir, "50m", "bec.tif"), 
  file.path(outdir, "50m", "modules", "landscape_binned.tif")
)

ggplot2::ggplot(check_bgc_out, ggplot2::aes(check_bgc_out$landscape)) +
      ggplot2::geom_histogram() +
      ggplot2::facet_wrap(~MAP_LABEL)

debugonce(check_roads)
check_roads(sf::st_read(file.path(outdir, "road_network.gpkg")))

roadsls <- sf::st_read(file.path(outdir, "road_network.gpkg"))
cities <- sf::st_read(file.path(outdir, "major_towns_bc.gpkg"))
nearest_town = "Hazelton"
start <- cities[cities$NAME == nearest_town,"NAME"]

debugonce(check_locations)
check_locations(roadsls, start)
```

gcperk commented on R/create_binned_landscape.R:
```diff
@@ -0,0 +1,72 @@
+#'   layers = c("dah_LS", "landform_LS","mrvbf_LS"))
+#' }
+create_binned_landscape <- function(
+    in_dir = fs::path(PEMprepr::read_fid()$dir_1020_covariates$path_rel, "25m", "modules_landscape"),
```
yes agree - this was an earlier hangover

gcperk commented on R/check_bgc_landscape.R:
```diff
@@ -0,0 +1,60 @@
+
+    routdf <- stats::na.omit(routdf)
+
+    ggplot2::ggplot(routdf, ggplot2::aes(routdf$landscape)) +
+      ggplot2::geom_histogram() +
+      ggplot2::facet_wrap(~MAP_LABEL)
```
I added a param with plot = T/F and then a print statement around the plot. I think this will work? or would it require a separate method?

ateucher replied:
Yup, that works to display the plot - a good option I think. It doesn't return the plot object, but I think that's ok if it's just for a quick look?

gcperk commented on R/check_bgc_landscape.R:
```diff
@@ -0,0 +1,60 @@
+#' @param binned_landscape A `SpatRast` with the landscape binned. This is the
+#'      output of the [create_binned_landscape()] function.
+#' @return A dataframe with summary of landsclass class and BEC unit.
+#'        A plot is also returned.
```
removed text and added plot = t/f param (see above). Not entirely sure if this will fix this entirely

gcperk commented on R/check_roads.R:
```diff
@@ -0,0 +1,71 @@
+  ## read in the major roads
+ # roads = sf::st_read(fs::path(PEMprepr::read_fid()$dir_1010_vector$path_rel,"road_network.gpkg"))
+
+  roads_check <- roads |>
```
Added check clause

gcperk commented on LICENSE:
Opps - did I? Not on pupose!

gcperk commented on R/check_bgc_cost.R:
```diff
@@ -0,0 +1,70 @@
+#'     fs::path(PEMprepr::read_fid()$dir_201010_inputs$path_abs,"acost.tif"))
+#' check_bgc_cost(bec, binned_landscape, cost)
+#' }
+check_bgc_cost <- function(bec, binned_landscape, cost) {
```
Yes - good point. I will update the name

gcperk commented on R/check_roads.R:
```diff
@@ -0,0 +1,71 @@
+    cli::cli_alert_success("check 2: road class does not contain overgrown road segment")
+  }
+
+  rSpd <- dplyr::tibble(
+    "road_surface" = c("resource", "unclassified", "recreation", "trail", "local", "collector", "highway", "service", "arterial", "freeway", "strata", "lane", "private", "yield", "ramp", "restricted", "water", "boat", "ferry", "driveway", "unclassifed"),
+    # "speed" = c(30, 30, 50, 4.5, 50, 80, 80, 50, 80, 80, 30, 30, 4.5, 30, 60, 4.5, 0.1, 0.1))
+    "speed" = c(3000, 3000, 5000, 4.5, 5000, 8000, 8000, 50, 8000, 8000, 3000, 3000, 4.5, 3000, 6000, 4.5, 0.1, 0.1, 3000, 4.5, 3000)
+  )
+
+  rclass <- data.table::as.data.table(unique(roads_check$road_surface))
+  names(rclass) <- "road_surface"
+  rdsAll <- merge(rclass, rSpd, by = "road_surface", all = T)
+  rdsAll <- rdsAll |>
+    dplyr::filter(is.na("speed"))
+  rdsAll <- as.data.frame(rdsAll)
```
There are some occasions where the user will want to manually edit the roads layer or compile different types of data not included in the bcdata download. This is a check to make sure everything will work in the cost layer. I will leave it as is, but flag it as a separate issue to review if actually needed.

gcperk commented on R/check_bgc_cost.R:
```diff
@@ -0,0 +1,70 @@
+      ))
+
+    return(rcdf_class)
+  }
```
Incorporated the changes in case_when, and updated the change to .default.

gcperk commented on R/check_roads.R:
```diff
@@ -0,0 +1,71 @@
+    sf::st_drop_geometry() |>
+    dplyr::select("ROAD_CLASS", "ROAD_SURFACE", "ROAD_NAME_FULL")
+
+  if (length(roads_check == 3)) {
```
issue #6

gcperk replied:
Changed this to check the required field names were included in the roads dataset

gcperk commented on R/create_cost_exclusion.R:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  if (fs::file_exists(fs::path(vec_dir, "water.gpkg"))) {
+    water <- sf::st_read(file.path(vec_dir, "water.gpkg")) |>
```
Added (just in case!)

gcperk commented on R/create_cost_exclusion.R:
```diff
@@ -0,0 +1,102 @@
+  if (fs::file_exists(fs::path(vec_dir, "road_network.gpkg"))) {
+    roads <- sf::st_read(file.path(vec_dir, "road_network.gpkg")) |>
+      sf::st_transform(3005)
+    roads_buff <- sf::st_buffer(roads, dist = 175)
```
Yes

gcperk commented on R/create_cost_penalty.R:
```diff
@@ -0,0 +1,236 @@
+      )
+    }
+
+    terra::writeRaster(hc_out, fs::path(output_file), overwrite = TRUE)
```
Yes - good call, added this as param when write_output = TRUE

gcperk commented on R/create_cost_penalty.R:
```diff
@@ -0,0 +1,236 @@
+    maxval <- qq$X90.
+  }
+
+  # 1. Assign high cost for cutblocks
```
Yes! Added

gcperk commented on R/create_cost_penalty.R:
```diff
@@ -0,0 +1,236 @@
+
+.assign_highcost <- function(shape, crs = 3005, costval, cost) {
+  hcsf <- sf::st_read(shape, quiet = TRUE) |>
+    sf::st_set_crs(crs) |>
```
These should all be 3005 or at least the default as read in by bcdata packages. This was a hold over from an earlier coding, that might not be relevant in this workflow. Updated to st_tranform()

gcperk commented:
@ateucher  - great suggestions here. FYI I have been updating the [vignettes](https://github.com/ninoxconsulting/PEMr/blob/main/vignettes/Setup-new-aoi.Rmd) when updating the PRs to try and make it easy to replicate when you want to do testing. There is a placeholder stand alone [R script](https://github.com/ninoxconsulting/PEMsamplr/blob/landscape_covars_create/R/placeholder_prep_cost_layer_old.R) to create a cost layer ( use the old one!). I can also upload a temp one in the ext folder on a temp basis?                                                             
                                                                                                                                                                                                                 I have incorporated all your suggested changes, excepted where noted and added the new checks. I still have a few remaining non visible binding for the plot_clhs_cost function, however I plan to rewrite this and drop the foreach %do% structure and change to purrr or something equivalent, so keeping them in for now. I will merge this one. Thanks for the detailed review !
